common/src/main/java/io/netty/util/concurrent/SingleThreadEventExecutor.java

```diff
@@ -979,81 +1093,127 @@ private boolean ensureThreadStarted(int oldState) {
                         }
                     }
-
                     try {
-                        // Run all remaining tasks and shutdown hooks. At this point the event loop
-                        // is in ST_SHUTTING_DOWN state still accepting tasks which is needed for
-                        // graceful shutdown with quietPeriod.
-                        for (;;) {
-                            if (confirmShutdown()) {
-                                break;
+                        if (shutdown) {
+                            // Run all remaining tasks and shutdown hooks. At this point the event loop
+                            // is in ST_SHUTTING_DOWN state still accepting tasks which is needed for
+                            // graceful shutdown with quietPeriod.
+                            for (;;) {
+                                if (confirmShutdown()) {
+                                    break;
+                                }
                             }
-                        }
 
-                        // Now we want to make sure no more tasks can be added from this point. This is
-                        // achieved by switching the state. Any new tasks beyond this point will be rejected.
-                        for (;;) {
-                            int oldState = state;
-                            if (oldState >= ST_SHUTDOWN || STATE_UPDATER.compareAndSet(
-                                    SingleThreadEventExecutor.this, oldState, ST_SHUTDOWN)) {
-                                break;
+                            // Now we want to make sure no more tasks can be added from this point. This is
+                            // achieved by switching the state. Any new tasks beyond this point will be rejected.
+                            for (;;) {
+                                int currentState = state;
+                                if (currentState >= ST_SHUTDOWN || STATE_UPDATER.compareAndSet(
+                                        SingleThreadEventExecutor.this, currentState, ST_SHUTDOWN)) {
+                                    break;
+                                }
                             }
-                        }
 
-                        // We have the final set of tasks in the queue now, no more can be added, run all remaining.
-                        // No need to loop here, this is the final pass.
-                        confirmShutdown();
+                            // We have the final set of tasks in the queue now, no more can be added, run all remaining.
+                            // No need to loop here, this is the final pass.
+                            confirmShutdown();
+                        } else {
+                            for (;;) {
+                                if (!runAllTasks()) {
+                                    break;
+                                }
+                            }
```

Comment:
I'm still confused by this `runnAllTasks()` loop: what tasks should we be running if we're suspending as part of the condition for suspending was that there were no tasks. If tasks were added after we made that decision then it should be up to the new thread to complete them.

Reply:
we could let the new thread handle it... I just thought we might want to drain as many as possible before hand over stuff to the new thread.

Reply:
that said it might be better to let this thread die as fast as possible if we already started a new one.